docs/src/what_is_aleph.md

```diff
@@ -0,0 +1,207 @@
+		if can_create(r):
+			Create a unit U of round r:
+				Pick as its parents at least N-f units (from different creators) of round (r-1), including node k previous unit.
+				Poll the input stream for a fresh data item `d` and place it in `U` in the `data` field
+			Send U to all other nodes
+			sleep(CREATE_DELAY)
+			break the inner loop
+```
+
+The `creator_task` keeps creating units of growing rounds, with the requirement that, roughly, there at least  `(2/3)*N` units in the previous round available in `D`. There is a certain delay after each unit is created -- this technicality allows us to make sure the Dag does not grow too quickly, which is problematic for pragmatic reasons: the nodes have limited bandwidth (which also varies between different nodes) and have bounded RAM memory, where we would like to store the Dag.
+
+
+#### 2.2.3  Malicious Units: Forks (Equivocations).
+Note that from the pseudocode of the `creator_task` it follows that each node is supposed to create exactly one unit per round, so ideally a Dag of "height" `R` should have `R*N` units in total. In the version of the Aleph protocol that is described in the main body of the [paper](https://arxiv.org/abs/1908.05156), this rule (one unit per node, per round) is enforced via a special broadcast protocol: Reliable Broadcast for disseminating units among other nodes. Roughly speaking, it guarantees that there will always be at most one unit per creator, per round. This is extremely convenient for making the protocol simple and clean, but unfortunately implementing the protocol this way (with reliable broadcast, which is quite heavy) is rather inefficient. For this reason Rush implements a version of the protocol (referred to as *QuickAleph* in the [paper](https://arxiv.org/abs/1908.05156)) that "allows" malicious nodes to create multiple versions of a unit for a given round (we call these forks). This makes the protocol slightly more complex, but much more efficient in practice. One can think of QuickAleph as an "optimistic" version of Aleph that assumes that all nodes typically behave honestly and fork only in extremely rare cases, at which time the protocol, in a sense, falls back to Reliable Broadcast in order to alert all the nodes that a forker has been detected (we give a little more details on that in a later section).
+
+
+#### 2.2.4  Disseminating Units.
+From now on we assume that whenever a unit `U` lands in a Dag `D` of an honest node `k` then all other honest nodes will eventually (maybe after some delay) receive `U` and place it in their copies of the Dag. To achieve this in practice there are several mechanisms in Rush to guarantee robustness of the process of disseminating units. First of all the creator broadcasts the unit several times (with exponentially increasing delays). Secondly, there is a request-response mechanism that allows to fetch missing units from other nodes.
+
+### 2.3 Computing the Ordering from Dag.
```

Comment:
This section is extremely heavy and difficult, and the recursive definition of how the nodes are ordered is quite hard to follow (at least to me). Maybe before you give the pseudoce listings it would be a good thing to add a high-level description of what is going to happen in the algorithm:

1. For every unit `U` we have the rules for virtual voting, which means that each element of `below(U)` is assigned a value of type `Option<bool>`.
2. If among the parents of a node `V` (which is beyond `U`), there is a quorum of `threshold` nodes for who voted `Some(true)` for `U`, then `V` witnesses that `U` is a valid candidate for the `Head` of the current round.
3. For each round `r = 0, ...`, as soon as some unit `U` of round `r` becomes a valid candidate for the `Head`, add all units from `below(U)` to the order (except those which are already there).

I am pretty sure that in the description above i oversimplified something to the point where it is no longer a valid description of the algorithm, but thi is the impression I got from reading it, and if I got something wrong, then likely the other readers will as well.

Reply:
Added a discussion just before Section 2.3.1.

But in general I don't want to repeat everything from the paper, this is supposed to be a TLDR for people who understand code better than math :)